### ISSUE TITLE
Correct sort line chart when the dimension axis has sort property

### ISSUE THREAD

**Issue by kanitw** [vega/vega-lite#1520]
For example, see this spec: http://pastebin.com/ZhAtaNw5.

The line chart doesn't suppose to be squiggled.

**Comment by kanitw** [comment]
Fixing this for raw plot is pretty straightforward.  
However, for aggregate plot this can be tricky as they require adding more fields to the aggregate transform to make sure we have the required field.

**Comment by kanitw** [comment]
> for aggregate plot, this can be tricky as they require adding more fields to the aggregate transform to make sure we have the required field.

To be precise, this might even require producing a separate data source.  

@jheer @arvind  Can we read aggregate data for scale sorting in Vega 3? 

(For example, if we sort an ordinal x-scale with "mean of A", we need to sort line value by mean of A too.  If we can't read the aggregate value that the scale already calculate, then we have to manually do it separately to sort line.)

**Comment by aldanor** [comment]
Just stumbled on this -- wonder if this was actually fixed?

Related: https://github.com/altair-viz/altair/issues/961

**Comment by JoeGermuska** [comment]
I believe I've just run into a related problem. When following the [Simple Bar Chart with Labels](https://altair-viz.github.io/gallery/bar_chart_with_labels.html) example in the Altair documentation, I found that adding the labels lost the ordering for the bar chart.

```python
bars = alt.Chart(soy).mark_bar(
             ).encode(y=alt.Y('State', 
                              sort=alt.EncodingSortField(field='Value', 
                                                         op='sum', 
                                                         order='descending')),
                      x = alt.X('Value', 
                                axis=alt.Axis(title='2017 Production in $')))
bars
```
![image](https://user-images.githubusercontent.com/46313/45855378-7df57000-bd15-11e8-8a05-38be1c21eed7.png)

```python
text = bars.mark_text(
    align='left',
    baseline='middle',
    dx=3,
).encode(
    text='Value',
)

bars + text
```
![image](https://user-images.githubusercontent.com/46313/45855409-9d8c9880-bd15-11e8-83ff-081711469d9a.png)

Hopefully this is helpful.

**Comment by domoritz** [comment]
No, this is related to layering and a limitation in Vega. We will address a subset of this issue I'm Vega-Lite 3.

**Comment by JoeGermuska** [comment]
@domoritz Thanks for clarifying.

**Comment by domoritz** [comment]
Do we have a smaller example? I'm not sure whether this is right or wrong. 

![image](https://user-images.githubusercontent.com/589034/46379658-617dff80-c654-11e8-9c4d-bae47ff6b569.png)

**Comment by aldanor** [comment]
@domoritz See the smaller example in https://github.com/altair-viz/altair/issues/961

**Comment by domoritz** [comment]
That's a great example. We should definitely fix this. 

[Open in the editor](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAkDODGALApgWwIaQFxUQFzwAdYsB6UgN2QHN0A6agSz0QFcAjOxge1IRQyUa6ALQAbZskoBmOgCY6ARjoArWNwB2kADTgoAE3R5MOUBAiQK6Ma2SxsYANp7zYYJABy6VMgeQACsh4yABOOlAAyvDcIb440gC+uq4Q7l4+fv7orGLhkFExcWCKSS7mad5FkACy6CEAnnkFsQ4ArAllALp6pRYYIQDWfhIavslQyBrR+owa1A5m5pAAHgtlFnj1hFUa3Kiz1jrrUABmjMhi+n7pY8eQ6iF4Do4BQaF5tQ15WTmQ3a69JaNUyQTbbPwAR1Y6A0eGYRkYVDyZwuVxw+WiLQ6EA6CSAA). 

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v3.2.1.json",
  "data": {
    "values": [
      {"Name": "Peter", "Score": 3},
      {"Name": "Paul", "Score": 1},
      {"Name": "Mary", "Score": 5}
    ]
  },
  "mark": "line",
  "encoding": {
    "x": {
      "type": "nominal",
      "field": "Name",
      "sort": ["Peter", "Mary", "Paul"]
    },
    "y": {"type": "quantitative", "field": "Score"}
  }
}
```

**Comment by domoritz** [comment]
Hmm, in this chart we should use the new sort index (`x_Name_sort_index`) instead of the name.

This is the wrong line:

```
"sort": {"field": "datum[\"Name\"]", "order": "descending"},
```

**Comment by domoritz** [comment]
Just to clarify, the behavior is not completely wrong since the order of the axis is not necessarily the same order as the points in the line. You can override the sort order. However, I think the default sort order should mirror the x order, though.

**Comment by aldanor** [comment]
> Just to clarify, the behavior is not completely wrong since the order of the axis is not necessarily the same order as the points in the line. You can override the sort order. However, I think the default sort order should mirror the x order, though.

@domoritz Yes, you can't say it's *wrong* per se, but there's two distinct options and you can't have them both as a default, that was my point exactly :) Out of the two, I agree, the default sort should mirror the x order as it would be least surprising in most (simple) cases.

**Comment by kanitw** [comment]
The small example case will be fixed in https://github.com/vega/vega-lite/pull/5089.  But we still have the aggregate plot case unfixed. I think that case is P3 though.

**Comment by devincody** [comment]
I'm running into a very similar case with Rule marks. As a minimal example based on https://github.com/vega/altair/issues/961

```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "data": {
    "values": [
      {"Name":"Peter", "Score": 3, "Name2":"Paul", "Score2": 4},
      {"Name":"Paul", "Score": 1, "Name2":"Paul", "Score2": 3},
      {"Name":"Mary", "Score": 5, "Name2":"Paul", "Score2": 4.5}
    ]
  },
  "mark": "rule",
  "encoding": {
    "x": {
      "type": "nominal",
      "field": "Name",
      "sort": ["Peter", "Mary",  "Paul"]
    },
    "x2": {"field":"Name2"},
    "y": {"type": "quantitative", "field": "Score"},
    "y2": {"field": "Score2"}
  }
}
```

<img width="58" alt="Image" src="https://github.com/user-attachments/assets/1d4b6100-e6cb-42e0-84db-d85b2ab6361a" />

As with the other examples, the X axis should be Peter, Mary, Paul, but isn't. 

I tried following the example of https://github.com/vega/vega-lite/pull/1523 to extend the solution to Rules ([see commit](https://github.com/devincody/vega-lite/commit/dd1e3d7f878f60c8c4c09f7349f487b7f476984c)), but couldn't get the axis to sort properly in my testing. Let me know if I should split this off into a separate issue.
